account_invoice_mode_monthly/models/sale_order.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from datetime import datetime
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, models
+from odoo.osv.expression import OR
+
+from odoo.addons.queue_job.job import job
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.model
+    def cron_generate_montly_invoices(self):
+        """Cron called daily to check if monthly invoicing needs to be done."""
+        company_ids = self._company_monthly_invoicing_today()
+        if company_ids:
+            self.generate_monthly_invoices(company_ids)
+
+    @api.model
+    def generate_monthly_invoices(self, companies=None):
+        """Generate monthly invoices for customers who require that mode."""
+        if not companies:
+            companies = self.company_id
+        partner_ids = self.read_group(
+            [
+                ("invoicing_mode", "=", "monthly"),
+                ("invoice_status", "=", "to invoice"),
+                ("company_id", "in", companies.ids),
+            ],
+            ["partner_invoice_id"],
+            groupby=["partner_invoice_id"],
+        )
+        for partner in partner_ids:
+            self.with_delay()._generate_invoices_by_partner(
+                partner["partner_invoice_id"][0]
+            )
+        companies.write({"invoicing_mode_monthly_last_execution": datetime.now()})
+        return partner_ids
+
+    @job(default_channel="root.invoice_monthly")
+    def _generate_invoices_by_partner(self, partner_id, invoicing_mode="monthly"):
+        """Generate invoices for a customer sales order."""
+        partner = self.env["res.partner"].browse(partner_id)
+        if partner.invoicing_mode != invoicing_mode:
+            return "Customer {} is not configured for monthly invoicing.".format(
+                partner.name
+            )
+        sales = self.search(
+            [
+                ("invoice_status", "=", "to invoice"),
+                ("partner_invoice_id", "=", partner.id),
+                ("order_line.qty_to_invoice", ">", 0),
+            ]
+        )
+        # By default grouped by partner/currency. Refund are not generated
+        invoices = sales._create_invoices(grouped=partner.one_invoice_per_order)
```

Comment:
Do not group sales having different Payment Term or Payment Mode
It would have been nice to be able to set a specific SO as "one_invoice_per_order"

Reply:
@jbaudoux I changed the machinery so sales orders with different payment terms are not grouped into the same invoice.
This is done before the creation of the jobs (by the cron). And can be easily extended.
Also for the payment mode this part of an other OCA module so will not be done here